cpp/serve/engine_actions/batch_verify.cc

```diff
@@ -0,0 +1,403 @@
+                            std::vector<int>& draft_output_tokens,
+                            std::vector<float>& draft_output_token_prob,
+                            std::vector<NDArray>& draft_output_prob_dist) {
+    int accept_length = 0;
+    RequestState rstate = estate->GetRequestState(request);
+    RequestModelState verify_mstate = rstate->mstates[verify_model_id_];
+    RequestModelState draft_mstate = rstate->mstates[draft_model_id_];
+    if (!draft_output_tokens.size()) {
+      // no draft tokens
+      // sample a new token
+      NDArray new_prob_dist = NDArray::Empty({1, ndata}, {kDLFloat, 32, 1}, {kDLCPU, 0});
+      new_prob_dist.CopyFromBytes(&p_probs[0], ndata * sizeof(float));
+      std::vector<int32_t> next_tokens = sampler_->SampleTokensFromProbs(
+          new_prob_dist, {verify_mstate}, {request->generation_cfg});
+      ICHECK(next_tokens.size() == 1U);
+      AcceptToken(estate, rstate, next_tokens[0]);
+      return 0;
+    }
+
+    ICHECK(verify_end - verify_start == (int)draft_output_tokens.size());
+    for (int i = verify_start; i < verify_end; ++i) {
+      int cur_token_idx = i - verify_start;
+      int cur_token = draft_output_tokens[cur_token_idx];
+      float p_value = p_probs[i * ndata + cur_token];
+      float q_value = draft_output_token_prob[cur_token_idx];
+      const float eps = 1e-9;
+      if (p_value >= q_value) {
+        AcceptToken(estate, rstate, cur_token);
+      } else {
+        float r = rng_.GetRandomNumber();
+        if (r < p_value / (q_value + eps)) {
+          AcceptToken(estate, rstate, cur_token);
+        } else {
+          // nomarlize a new probability distribution
+          double sum_v = 0.0;
+          std::vector<float> new_prob;
+          new_prob.resize(ndata);
+
+          NDArray q_dist = draft_output_prob_dist[cur_token_idx];
+          if (q_dist->device.device_type != kDLCPU) {
+            q_dist = q_dist.CopyTo(DLDevice{kDLCPU, 0});
+          }
+
+          ICHECK(q_dist->device.device_type == kDLCPU);
+
+          ICHECK(q_dist->ndim == 1);
+          ICHECK(ndata == q_dist->shape[q_dist->ndim - 1]);
+          const float* __restrict p_qdist =
+              static_cast<float*>(__builtin_assume_aligned(q_dist->data, 4));
+
+          // this part is parallelizable
+          // TODO: parallelize this part
+          for (int j = 0; j < ndata; ++j) {
+            new_prob[j] = p_probs[i * ndata + j] - p_qdist[j];
+            new_prob[j] = new_prob[j] < 0 ? 0 : new_prob[j];
+            // new_prob[j] = std::exp(new_prob[j]);
+            sum_v += new_prob[j];
+          }
+          for (int j = 0; j < ndata; ++j) {
+            new_prob[j] /= sum_v;
+          }
+
+          // sample a new token from the new distribution
+          NDArray new_prob_dist = NDArray::Empty({1, ndata}, {kDLFloat, 32, 1}, {kDLCPU, 0});
+          new_prob_dist.CopyFromBytes(new_prob.data(), ndata * sizeof(float));
```

Comment:
Directly write normalized probability to `p_probs` and avoid allocation. The pre-normalized p_prob distribution will be no longer used.

Reply:
Fixed. But still need allocating some new buffers. Wondering if it is possible to achieve zero allocation here?